### PR TITLE
Bugfix: Make `doOnce` compatible with `event.isRegistered` and `event.unregister`

### PR DESCRIPTION
**Pull request by herbertgithubaccount**
Previously if `doOnce` was passed to `event.register`, then a new anonymous function was created, and then that function was registered. Here is a minimum code sample to reproduce the issue:
```lua
local function func()
    mwse.log("func callback")
end
event.register("load", func, {doOnce=true})
mwse.log("is func registered? %s", event.isRegistered("load", func)) --> `false`
```

This was not explained in the documentation and led to some weird behavior.

My solution was to create a conversion table that converts the `originalCallback` to the `newCallback`. (Where the `newCallback` is a wrapper that calls `originalCallback` and then unregisters it.) Whenever `event.unregister` is called on a `doOnce` function, the `originalCallback` is used to retrieve the `newCallback`, and then the relevant entry is purged from the conversions table, and the `newCallback` is unregistered.

The relevant entry from the conversion table is also purged from the `newCallback` itself, to minimize errors. More specifically, the `newCallback` is defined as follows:
```lua
local originalCallback = callback
local function newCallback(e)
    -- `event.unregister` will handle the conversion for us.
    this.unregister(eventType, originalCallback, options)
    originalCallback(e)
end
```

I did some testing, and found that `event.unregister`, and `event.isRegistered` now work properly with `doOnce` callbacks. And the callbacks still get unregistered after being called once. The only change is that they can now be unregistered earlier if deisred.

### CODE REVIEW

File: misc/package/Data Files/MWSE/core/lib/event.lua [hunk 1]
```diff
@@ -93,10 +97,14 @@ function this.register(eventType, callback, options)
 	-- If 'doOnce' was set, wrap with a call to unregister.
 	if options.doOnce then
 		local originalCallback = callback
-		callback = function(e)
-			this.unregister(eventType, callback, options)
+		local function newCallback(e)
+			-- `event.unregister` will handle the conversion for us.
+			event.unregister(eventType, originalCallback, options)
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/lib/event.lua:102]: Let's just call this `this.unregister` to stay consistent with the rest of the file.

**Review by C3pa — COMMENTED** [review]
Other than the small remark below, this works as expected.

File: misc/package/Data Files/MWSE/core/lib/event.lua (hunk 1 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/lib/event.lua]: Good catch. I fixed this in the latest commit.

**Comment by herbertgithubaccount** [comment]
I converted this to a draft because I believe I found a problem with it.

The `local found = table.find(callbacks, callback)` does not work as expected when `doOnce == true`. And this can result in the `newCallback` not being properly unregistered in the case of duplicate registration.

I'll get started on a fix.

**Comment by C3pa** [comment]
Well, if there are some edge cases, you can add some unit tests to cover them. We have tests in https://github.com/MWSE/MWSE/tree/master/misc/package/Data%20Files/MWSE/core/core%20tests.

File: misc/package/Data Files/MWSE/core/lib/event.lua [hunk 2]
```diff
@@ -76,45 +78,47 @@ local function remapFilter(options, showWarnings)
...
 
+
+local doOnceCallbacks = {} ---@type table<fun(...):..., fun(...):...>
+
+
 function this.register(eventType, callback, options)
-	-- Validate event type.
-	if (type(eventType) ~= "string" or eventType == "") then
-		return error("event.register: Event type must be a valid string.")
-	end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/lib/event.lua:83]: Why did you remove this check?

File: misc/package/Data Files/MWSE/core/lib/event.lua [hunk 3]
```diff
@@ -76,45 +78,47 @@ local function remapFilter(options, showWarnings)
...
 
-	-- Validate callback.
-	if (type(callback) ~= "function") then
-		return error("event.register: Event callback must be a function.")
-	end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/lib/event.lua:88]: Same as above

File: misc/package/Data Files/MWSE/core/lib/event.lua (hunk 2 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/lib/event.lua:83]: I removed those checks because they're happening in `event.isRegistered` and the newer implementation calls `event.isRegistered` from inside `event.register`.

**Comment by herbertgithubaccount** [comment]
Adding test cases would be a good idea. I'll try to do that soon. It's been a surprisingly busy month but hopefully I can find some time soon.

**Comment by C3pa** [comment]
> It's been a surprisingly busy month but hopefully I can find some time soon

No rush Herbert, this is just a hobby! I hope you are doing well.

File: misc/package/Data Files/MWSE/core/lib/event.lua (hunk 2 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/lib/event.lua]: Very good point about the error messsages. I did not consider that in my original approach. I reverted that change and also tried in general to refactor this PR so that it contains minimal code changes with upstream/master. I'll add in some tests later today.

File: misc/package/Data Files/MWSE/core/lib/event.lua [hunk 4]
```diff
@@ -93,12 +98,17 @@ function this.register(eventType, callback, options)
 	-- If 'doOnce' was set, wrap with a call to unregister.
 	if options.doOnce then
 		local originalCallback = callback
```
Comment (herbertgithubaccount) [inline misc/package/Data Files/MWSE/core/lib/event.lua:100]: I likely don't need to have three different callback variables (`callback`, `originalCallback`, and `newCallback`) floating around here, but I did it this way for maximum clarity surrounding the definition of `newCallback`.

This was done because 
```lua
local function func()
...
end
```
and 
```lua
local func = function()
...
end
```
behave differently with respect to what the variable `func` resolves to in the body of the function. So the use of  `oldCallback` and `newCallback` allow us to avoid that complexity and have more confidence about which functions are being executed where.

**Comment by herbertgithubaccount** [comment]
I merged the changes with the master branch and added in some test coverage to the `doOnce` functionality. Everything seems to work as expected, but let me know if there is something I missed.

File: misc/package/Data Files/MWSE/core/core tests/tests/events.lua [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+-- 	if not name then break end
+-- 	if name == "generalEvents" then
+-- 		generalEvents = value
+-- 	elseif name == "filteredEvents" then
+-- 		filteredEvents = value
+-- 	end
+-- end
+
+testSuite:test("Load generalEvents and filteredEvents", function ()
+	-- testSuite.logger:debug("checking if generalEvents was loaded")
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/core tests/tests/events.lua:104]: Can you delete commented-out code?

File: misc/package/Data Files/MWSE/core/core tests/main.lua [hunk 6]
```diff
@@ -9,6 +9,7 @@ end
 if (table.empty(tests)) then
 	tests = {
 		"dialogueReplacement",
+		"events.lua",
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/core tests/main.lua:12]: The files listed are without the `.lua` extension since these are fed into `dofile`.

File: misc/package/Data Files/MWSE/core/core tests/tests/events.lua [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+		testSuite:expect(isRegistered(callback1)).toBe(true)
+		testSuite:expect(isRegistered(callback2)).toBe(true)
+		testSuite:expect(isRegistered(callback3)).toBe(true)
+
+		testSuite:expect(innerCallbacks.callback1).toBeCalledTimes(2)
+
+	end
+	
+	-- unregister the first callback and see if the others behave normally.
+	unregister(callback1)
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/core tests/tests/events.lua:267]: It's a bit unintuitive to use `unregister` because it passes `doOnce = true`, but the `callback1` wasn't registered that way.

File: misc/package/Data Files/MWSE/core/lib/event.lua (hunk 4 above)
Comment (C3pa) [reply misc/package/Data Files/MWSE/core/lib/event.lua:100]: The changes look good.

**Review by C3pa — CHANGES_REQUESTED** [review]

File: misc/package/Data Files/MWSE/core/core tests/tests/events.lua (hunk 7 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/core tests/tests/events.lua]: That's a good point. I've made the change.

File: misc/package/Data Files/MWSE/core/core tests/main.lua (hunk 6 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/core tests/main.lua]: Oh oops. Good catch. Sorry about that.

File: misc/package/Data Files/MWSE/core/core tests/tests/events.lua (hunk 5 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/core tests/tests/events.lua]: It has been done. Sorry, I was a bit sloppy with the comments in that file.

**Comment by C3pa** [comment]
Thanks!
